core/network-libp2p/src/config.rs

```diff
@@ -0,0 +1,318 @@
+					secp256k1::SecretKey::generate)
+				.map(secp256k1::Keypair::from)
+				.map(Keypair::Secp256k1),
+
+			Ed25519(Secret::New) =>
+				Ok(Keypair::generate_ed25519()),
+
+			Ed25519(Secret::Input(k)) =>
+				Ok(Keypair::Ed25519(k.into())),
+
+			Ed25519(Secret::File(f)) =>
+				get_secret(f,
+					|mut b| ed25519::SecretKey::from_bytes(&mut b),
+					ed25519::SecretKey::generate)
+				.map(ed25519::Keypair::from)
+				.map(Keypair::Ed25519),
+		}
+	}
+}
+
+/// Load a secret key from a file, if it exists, or generate a
+/// new secret key and write it to that file. In either case,
+/// the secret key is returned.
+fn get_secret<P, F, G, E, K>(file: P, parse: F, generate: G) -> io::Result<K>
+where
+	P: AsRef<Path>,
+	F: for<'r> FnOnce(&'r mut [u8]) -> Result<K, E>,
+	G: FnOnce() -> K,
+	E: Error + Send + Sync + 'static,
+	K: AsRef<[u8]>
+{
+	std::fs::read(&file)
+		.and_then(|mut sk_bytes|
+			parse(&mut sk_bytes)
+				.map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e)))
+		.or_else(|e| {
+			if e.kind() == io::ErrorKind::NotFound {
+				file.as_ref().parent().map_or(Ok(()), fs::create_dir_all)?;
+				let sk = generate();
+				write_secret_file(file, sk.as_ref())?;
+				Ok(sk)
+			} else {
+				Err(e)
+			}
+		})
+}
+
+/// Write secret bytes to a file.
+fn write_secret_file<P>(path: P, sk_bytes: &[u8]) -> io::Result<()>
+where
+	P: AsRef<Path>
+{
+	let mut file = open_secret_file(&path)?;
+	file.write_all(sk_bytes)
+}
+
+/// Opens a file containing a secret key in write mode.
+#[cfg(unix)]
+fn open_secret_file<P>(path: P) -> io::Result<fs::File>
+where
+	P: AsRef<Path>
+{
+	use std::os::unix::fs::OpenOptionsExt;
+	fs::OpenOptions::new()
+		.write(true)
+		.create_new(true)
+		.mode(0o600)
+		.open(path)
+}
+
+/// Opens a file containing a secret key in write mode.
+#[cfg(not(unix))]
+fn open_secret_file<P>(path: P) -> Result<fs::File, io::Error>
+where
+	P: AsRef<Path>
+{
+	fs::OpenOptions::new()
+		.write(true)
+		.create_new(true)
+		.open(path)
+}
+
+#[cfg(test)]
+mod tests {
+	use quickcheck::*;
+	use rand::distributions::*;
+	use rand::Rng;
+	use super::*;
+	use tempdir::TempDir;
+
+	#[derive(Debug, Clone)]
+	struct AlnumString(String);
+
+	impl Arbitrary for AlnumString {
+		fn arbitrary<G: Gen>(mut g: &mut G) -> AlnumString {
+			let n = g.gen_range(1, g.size() + 1);
+			AlnumString(Alphanumeric.sample_iter(&mut g).take(n).collect())
+		}
+	}
+
+	fn secret_bytes(kp: &Keypair) -> Vec<u8> {
+		match kp {
+			Keypair::Ed25519(p) => p.secret().as_ref().iter().cloned().collect(),
+			Keypair::Secp256k1(p) => p.secret().as_ref().iter().cloned().collect(),
+			_ => panic!("Unexpected keypair.")
+		}
+	}
+
+	#[test]
+	fn test_secret_file_mkdir() {
+		fn prop(dirname: AlnumString, filename: AlnumString) -> io::Result<bool> {
+			let tmp = TempDir::new(&dirname.0)?;
+			std::fs::remove_dir(tmp.path())?;
+			let file = tmp.path().join(filename.0).to_path_buf();
+			NodeKeyConfig::Ed25519(Secret::File(file)).into_keypair().map(|_| true)
```

Comment:
You don't check that the directory is created here?

Reply:
If it wouldn't be, the test would fail because its trying to write a file into that directory.

Reply:
Yeah, but you should try to write tests as a blackbox. You want to make sure that the test passes, because the file was created. If I would alter the implementation to just return `Ok(())`, the test would still pass.

Reply:
> Yeah, but you should try to write tests as a blackbox. You want to make sure that the test passes, because the file was created. If I would alter the implementation to just return `Ok(())`, the test would still pass.

It would pass if the implementation would just return `Ok(some_random_keypair)`, that is true. Checking that the file exists is certainly better.  :+1: